modules/banner/scanner.go

```diff
@@ -0,0 +1,152 @@
+	"strings"
+
+	"github.com/zmap/zgrab2"
+)
+
+// Flags give the command-line flags for the banner module.
+type Flags struct {
+	zgrab2.BaseFlags
+	Probe    string `long:"probe" default:"\n" description:"Probe to send to the server."`
+	Pattern  string `long:"pattern" description:"Pattern to match, must be valid regexp."`
+	MaxTries int    `long:"maxtries" default:"1" description:"Number of tries for timeouts and connection errors before giving up."`
+}
+
+// Module is the implementation of the zgrab2.Module interface.
+type Module struct {
+}
+
+// Scanner is the implementation of the zgrab2.Scanner interface.
+type Scanner struct {
+	config *Flags
+	regex  *regexp.Regexp
+}
+
+type Results struct {
+	Banner string `json:"banner,omitempty"`
+	Length int    `json:"length,omitempty"`
+}
+
+// RegisterModule is called by modules/banner.go to register the scanner.
+func RegisterModule() {
+	var module Module
+	_, err := zgrab2.AddCommand("banner", "Banner", "Grab banner by sending probe and match with regexp", 80, &module)
+	if err != nil {
+		log.Fatal(err)
+	}
+}
+
+// NewFlags returns a new default flags object.
+func (m *Module) NewFlags() interface{} {
+	return new(Flags)
+}
+
+// GetName returns the Scanner name defined in the Flags.
+func (scanner *Scanner) GetName() string {
+	return scanner.config.Name
+}
+
+// GetTrigger returns the Trigger defined in the Flags.
+func (scanner *Scanner) GetTrigger() string {
+	return scanner.config.Trigger
+}
+
+// Protocol returns the protocol identifier of the scan.
+func (scanner *Scanner) Protocol() string {
+	return "banner"
+}
+
+// GetPort returns the port being scanned.
+func (scanner *Scanner) GetPort() uint {
+	return scanner.config.Port
+}
+
+// InitPerSender initializes the scanner for a given sender.
+func (scanner *Scanner) InitPerSender(senderID int) error {
+	return nil
+}
+
+// NewScanner returns a new Scanner object.
+func (m *Module) NewScanner() zgrab2.Scanner {
+	return new(Scanner)
+}
+
+// Validate validates the flags and returns nil on success.
+func (f *Flags) Validate(args []string) error {
+	return nil
+}
+
+// Help returns the module's help string.
+func (f *Flags) Help() string {
+	return ""
+}
+
+// Init initializes the Scanner with the command-line flags.
+func (scanner *Scanner) Init(flags zgrab2.ScanFlags) error {
+	f, _ := flags.(*Flags)
+	scanner.config = f
+	scanner.regex = regexp.MustCompile(scanner.config.Pattern)
+	return nil
+}
+
+var NoMatchError = errors.New("pattern did not match")
+
+func (scanner *Scanner) Scan(target zgrab2.ScanTarget) (zgrab2.ScanStatus, interface{}, error) {
+	try := 0
+	var (
+		conn    net.Conn
+		err     error
+		readerr error
+	)
+	for try < scanner.config.MaxTries {
+		try += 1
+		conn, err = target.Open(&scanner.config.BaseFlags)
+		if err != nil {
+			continue
+		}
+		break
+	}
+	if err != nil {
+		return zgrab2.TryGetScanStatus(err), nil, err
+	}
+	defer conn.Close()
+	r := strings.NewReplacer(`\n`, "\n", `\r`, "\r", `\t`, "\t")
```

Comment:
Any reason for replacing just these, and not using e.g. JSON or another standard special-characters-in-strings format? I could see wanting to send a hex string, for example.

Reply:
Sending hex string would make sense. Initially I was just using it to send weird HTTP requests. The problem is refection.tag.get will convert everything to literal string. I can't seem to find a good way around it. Any suggestions?

Reply:
Repacling \x to non-literal in the last commit.

Reply:
What you have is probably preferable to forcing hex input -- I was picturing something like `json.Unmarshal([]byte("\"" + scanner.config.Probe + "\""),  &scanner.config.Probe)`, so you could do e.g. `--probe="GET /\r\n\r\n"` or `--probe="GET /\u000A\u000D\u000A\u000D"`, but that may be more complicated than necessary.

The type of issue with the current implementation (as opposed to a standard escaped-string-decoding library call) is that e.g. `"\\n"` -- which one would expect to send a literal `\n` -- would here send a slash followed by a linebreak.

Reply:
the problem with unmarshalled json is it only accepts unicode encoded hex string such as \u0000 not \x00. I'll work on it from another angle maybe using regexp replace

Reply:
Alright, problem fixed. Although triple escape is a bit annoying but I like this solution much better.

Reply:
Would it possible to send binary data as a probe request? For example, [most] RDP services can be identified by sending and receiving a specific value. `echo 0300000b06e0000000000 | xxd -r -p | nc 192.168.1.2 3389`. Listening servers respond with `\x03\x00\x00\x0b\x06\xd0\x00\x00\x124\x00`